File: #79_Word_Search/Solution.java
public class Solution {
    public boolean exist(char[][] board, String word) {
        int rows = board.length;
        int cols = board[0].length;
        boolean[][] isVisited = new boolean[rows][cols];
        int count = 0; /* ptr for char in word */
        boolean hit =false;
        for (int i = 0; i < rows; i++) {
        	for (int j = 0; j < cols; j++) {
        		if (board[i][j] == word.charAt(0)) {
        			if (word.length() == 1) return true;
        			hit = searchBoard(board, isVisited, word, 0, i, j);
        			if (hit) return hit;
        		}
        	}
        }
        return hit;
    }

    private boolean searchBoard(char[][] board, boolean[][] isVisited, String word, int ptr, int r, int c) {
    	isVisited[r][c] = true;
    	ptr++;
    	int rows = board.length;
        int cols = board[0].length;
    	char search = word.charAt(ptr);
    	boolean ret =false;
    	if (isIndexValid(c - 1, cols) && board[r][c - 1] == search && isVisited[r][c - 1] == false) {
    		if (ptr == word.length() - 1) return true;
    		ret = searchBoard(board, isVisited, word, ptr, r, c - 1);
    		if (ret) return ret;
    	}
    	if (isIndexValid(c + 1, cols) && board[r][c + 1] == search && isVisited[r][c + 1] == false) {
    		if (ptr == word.length() - 1) return true;
    		ret = searchBoard(board, isVisited, word, ptr, r, c + 1);
    		if (ret) return ret;
    	}
    	if (isIndexValid(r - 1, rows) && board[r - 1][c] == search && isVisited[r - 1][c] == false) {
    		if (ptr == word.length() - 1) return true;
    		ret = searchBoard(board, isVisited, word, ptr, r - 1, c);
    		if (ret) return ret;
    	}
    	if (isIndexValid(r + 1, rows) && board[r + 1][c] == search && isVisited[r + 1][c] == false) {
    		if (ptr == word.length() - 1) return true;
    		ret = searchBoard(board, isVisited, word, ptr, r + 1, c);
    		if (ret) return ret;
    	}
    	isVisited[r][c] = false;
    	return ret;
    }

    private boolean isIndexValid(int index, int bound) {
    	return (index < bound) && (index >= 0);
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	char[][] board = {
    		{'A', 'B', 'C', 'E'},
    		{'S', 'F', 'C', 'S'},
    		{'A', 'D', 'E', 'E'}
    	};
    	System.out.println(sol.exist(board, "ABCCED"));
    	System.out.println(sol.exist(board, "SEE"));
    	System.out.println(sol.exist(board, "ABCB"));
    }
}

File: #79_Word_Search/Solution.py
class Solution:
    # @param {character[][]} board
    # @param {string} word
    # @return {boolean}
    def exist(self, board, word):
        if not board and not word:
        	return True
        if not board:
        	return False
        if not word:
        	return True
        for i in range(len(board)):
        	for j in range(len(board[0])):
        		if board[i][j] == word[0]:
        			if self.search(board, word, i, j):
        				return True
        return False

    def search(self, board, word, i, j):
    	if board[i][j] != word[0]:
    		return False
    	if len(word) == 1:
    		return True
    	board[i][j] = '0'
    	if i > 0 and board[i - 1][j] != '0' and self.search(board, word[1:], i - 1, j):
    		return True
    	if i < len(board) - 1 and board[i + 1][j] != '0' and self.search(board, word[1:], i + 1, j):
    		return True
    	if j > 0 and board[i][j - 1] != '0' and self.search(board, word[1:], i, j - 1):
    		return True
    	if j < len(board[0]) - 1 and board[i][j + 1] != '0' and self.search(board, word[1:], i, j + 1):
    		return True
    	board[i][j] = word[0]
    	return False

sol = Solution()
t = [['A', 'B', 'C', 'E'], ['S', 'F', 'C', 'S'], ['A', 'D', 'E', 'E']]
print sol.exist(t, "ABCB")